kep/4986-sakkara-cluster-topology/README.md

```diff
@@ -0,0 +1,536 @@
+Sakkara employs an open-source solver which uses the [chic-sched](https://github.com/ibm/chic-sched) algorithm to find a placement solution for the entire group, satisfying the group topological constraints. (Other solvers may be substituted.)
+
+## Motivation
+
+A requirement for long running jobs with a specific pattern of interaction among its component pods is topology-aware scheduling of the pods, given a cluster topology.
+Existing solutions are as follows.
+
+- [node resource topology](https://github.com/kubernetes-sigs/scheduler-plugins/tree/master/pkg/noderesourcetopology): concerned about the connectivity of resources within a node
+- [coscheduling](https://github.com/kubernetes-sigs/scheduler-plugins/tree/master/pkg/coscheduling): group placement without topology awareness, nor group preemption
+- [pod topology spread constraints](https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/): a very narrow form of topology constraint of a single pod
+
+Lacking is a scheduler that is both topology-aware at the cluster level, with a set of rich topological constraints, and places/preempts whole groups.
+
+### Goals
+
+Develop a scheduler plugin which places a group of pods in a cluster with a given tree topology, satisfying group placement constraints at all levels of the hierarchy.
+
+### Non-Goals
+
+- Discovery of cluster topology
+- Migration of pods
+- Elastic groups with changing size
+
+## Proposal
+
+### User Stories
+
+1- **User A:**
+
+Want to deploy a job with 8 pods, equally partitioned into two partitions, where each partition is placed in separate rack, and the pods belonging to a partition are packed on nodes within the rack. Partitioning improves availability, and packing improves performance.
+
+![before-1](images/use-cases/before-1.png)
+
+![after-1](images/use-cases/after-1.png)
+
+2- **User B:**
+
+Want to place a job with 6 pods packed on same physical server, as much as possible, where pods on the same server are spread across nodes in that server, but in the range [2,4]. Having all pods on the same physical server improves communication among the pods. And, having less than 2 pods in a node spreads the pods too thinly, and having more than 4 may leads to congestion.
+
+![before-2](images/use-cases/before-2.png)
+
+![after-2](images/use-cases/after-2.png)
+
+### Examples
+
+![topology](images/example/topology.png)
+
+![group](images/example/group.png)
+
+![placement](images/example/placement.png)
+
+![result-group](images/example/result-group.png)
+
+![result-pod](images/example/result-pod.png)
+
+![premption before](images/preemption/before.png)
+
+![premption after-1](images/preemption/after-1.png)
+
+![premption after-2](images/preemption/after-2.png)
+
+### Notes/Constraints/Caveats
+
+### Risks and Mitigations
+
+## Design Details
+
+![architecture](images/architecture.png)
+
+![plugin state transitions](images/state-transitions.png)
```

Comment:
After reading the Design section, I still cannot figure out the exact responsibility of each phase. Like, which phase does the "pre-scheduling" to update the pod's label, what if the "pre-scheduling" result is in conflict with the real-time scheduling decision, how preemption works, etc.

It's quite essential to illustrate: by given a group of pods carrying the group constraints, in an e2e manner how they're processed in each phase of a scheduling cycle.